### ISSUE TITLE
Add variant_sku and price

### ISSUE THREAD

**Issue by savrand** [autotelik/datashift_spree#29]
I have added to csv fields as it is indicated https://github.com/autotelik/datashift_spree/wiki/Spree-Import-Export-Tips#variants
variant_sku . before saving  
https://github.com/autotelik/datashift/blob/master/lib/loaders/loader_base.rb#L346
I output @load_object.variants and it shows me variants with filled  sku  as in csv but after saving they are changed to standard ones.

**Comment by chekovz** [comment]
This is still an issue.

I don't completely understand all the code so I'm not sure where to stand but from what I gather, the variant_sku are correctly assigned but not saved.

When the loop is done, @load_object.save is called, which attemps to save the master product only, so it seems like the Variant SKUs are getting replaced locally but the @load_object affects only the master and the changes are not being persisted.

Since they use the same code, I guess it affects variant_price and count_on_hand equally.

Any suggestions?

**Comment by autotelik** [comment]
Some changes to fix count_on_hand for Spree 2 have literally just been checked in.

Sorry bit hard to follow the trail ..can you specify exactly which fields are not being persisted as you expect .. On Variants ?

sku, price and count_on_hand ?  For what version of Spree ? And using master branch from github (not gem) ?

thanks
tom

**Comment by chekovz** [comment]
I've made some progress on the issue

I'm using the latest version of everything :

Datashift v0.16.0@d561aa (from github)
Datashift_spree v.0.6.0@d6e271

Spree 2.4.3

I've found the issue though :

SKU are not updated because the children (Spree::Variant) are updated locally through the parent (Spree::Product) and Spree::Product does not accepts_nested_attributes for Spree::Variant so the changes on the children are not persisted when saving the parent.

So it's possible to replace @load_object.save by          

@load_object.variants.each do |i|
    i.save
end

Or we could use v.update_attributes( sku: values[i].to_s") instead.

I don't know the project well enough to be know the best way to fix it, what are your thoughts?

(English is not my mother tongue so sorry if it's not exactly clear)

**Comment by redglory** [comment]
Hum... I've not changed any code regarding 'variant_sku' method and I got my custom variant sku's saved...

**Comment by dangerdogz** [comment]
I will post a more detailed use case later to do with a CSV sample, dependencies version, log entries from Datashift and resulting SQL.

**Comment by dangerdogz** [comment]
Sample CSV

| SKU      | Name         | ShippingCategory | Price    | Option Types    | Option Types    | Option Types     | variant_sku                |
|----------|--------------|------------------|----------|-----------------|-----------------|------------------|----------------------------|
| ILJR0247 | TSHIRT PROMO | DEFAUT           | 100 | color:Q; size:M | color:Q; size:G | color:A; size:TP | 10583856\|10583857\|10583858 |

Resultings SKU :

| ILJR0247_1
| ILJR0247_2
| ILJR0247_3

Variant_SKU is processed without problem but values aren't overwritten. From the logs, it's clear that after looping over @load_object. variants, the SKUs are replaced locally but the changes aren't persisted when the Product is saved.

If you look at a relevant part of the log, you see that no UPDATE is being called on any variant.

http://pastebin.com/uLLc8NTC

Could you post a sample of your working CSV ? Also, which version of Rails/Spree are you running?

I might be doing something wrong but still, if I save the Product.variants explicitly in the 'variant_sku' block, it works

I'm curious about your thoughts on this

**Comment by autotelik** [comment]
redglory  kindly commited some fixes all around this yesterday  (https://github.com/autotelik/datashift_spree/commit/936b96ef04994b1676ad9d5f602981edbef4d0f0)

Have you tried pulling this latest from master ? if this does not fix your problem just let me know and I'll try and go through your examples/logs.

**Comment by autotelik** [comment]
@load_object.variants.each_with_index {|v, i| v.sku = values[i].to_s }
@load_object.save 

I assume save on Parent product peculates down to associations but given what you see perhaps not

I don't have time to dev/test this right now, so if you're desperate ..try changing

Line 154 .... @load_object.variants.each_with_index {|v, i| v.sku = values[i].to_s }

To save the Variant itself, something like :

@load_object.variants.each_with_index do |v, i| 
v.sku = values[i].to_s 
begin
    v.save!
rescue => x
     logger.error "Warning SKU not set on Variant #{v.inspect} #{e.message}"
end

cheers
tom

**Comment by dangerdogz** [comment]
I will try again with the master branch and get back to you

Cheers

**Comment by joeswann** [comment]
Just ran into pretty much this issue with spree 3.0.0 and variant_price

Changing line 112 to this seemed to do the trick
    @load_object.variants.each_with_index {|v, i| v.update_attributes(price: values[i].to_f) }

**Comment by sybind** [comment]
@joeswann Thanks, that fixed it per your suggestion.
```
@load_object.variants.each_with_index {|v, i| v.update_attributes(sku: values[i].to_f) }
```
